avocado/core/test.py

```diff
@@ -133,6 +133,98 @@ def str_filesystem(self):
                                  % (self.str_uid, str(self)))
 
 
+def get_file_data_dir(test):
+    """
+    Returns the path to the file-level data directory
+
+    For test a test file hosted at /usr/share/avocado/tests/sleeptest.py
+    the file-level data dir is /usr/share/avocado/tests/sleeptest.py.data.
+
+    Note that this directory has no specific relation to the test
+    name, only to the file that contains the test.  It can be used to
+    host data files that are generic enough to be used for all tests
+    contained in a given test file.
+
+    This function is currently a thin wrapper (almost an alias) for
+    :attr:`avocado.core.test.Test.datadir`.  The idea is just to give
+    consistent naming across the various data dir, and prepare for the
+    removal of :attr:`avocado.core.test.Test.datadir`.
+
+    :param test: the test to locate its file-level data directory
+    """
+    # TODO: on removal of Test.datadir, move its code over here
+    return test.datadir
+
+
+def get_test_data_dir(test):
+    """
+    Returns the path to the test-level data directory
+
+    For an INSTRUMENTED test, implemented by a method named `test`, on
+    a class named `SleepTest`, hosted at a file located at
+    /usr/share/avocado/tests/sleeptest.py, the test-level data directory
+    is /usr/share/avocado/tests/sleeptest.py.data/SleepTest.test
+
+    For a SIMPLE test, the test is the file, so this is the same as
+    :func:`get_file_data_dir`.
+
+    :param test: the test to locate its file-level data directory
+    """
+    file_data_dir = get_file_data_dir(test)
+    if file_data_dir is None:
+        return None
+
+    if isinstance(test, SimpleTest):
```

Comment:
I don't like hardcoded exceptions, so let's talk about this one. Yes, the SimpleTest is not likely to contain multiple tests, but:

1. it won't break anything if `SimpleTest.test` is used
2. inherited classes might actually define multiple `test` methods

So I'd vote for systematic solution which might add not-really-needed paths, but without surprises.

Reply:
I share the same feeling with regards to "exceptions", or special treatment, like this one.  I initially chose this approach because it saved other classes from an added responsibility, that is, to declare the filename and/or datadir associated with them.

The option that I see is for a test class to know its test-level data dir, which I also find cumbersome.  It'd be something like:

```python
class SimpleTest(Test):
   _test_level_data_dir = datadir
```

That would work because of this:

```python
    @property
    def filename(self):
        """
        Returns the name of the file (path) that holds the current test
        """
        return os.path.abspath(self.name.name)
```

And because of the unchanged `datadir` property in `SimpleTest`.  The idea is generic enough, though: you delegate the responsibility to the class.  It means more code, but a more consistent and predictable solution.

Again, I chose the simpler solution for now, but I'd have no problems in moving to another one, like the one I exemplified now.

Reply:
My point is you don't actually need any special handling. The SimpleTest does contain class/method attributes, therefor it can generate the test-datadir. It'll always be the same, therefor not much useful, but it'll be available and consistent with other implementation.

Also I pointed out this should only be exception for SimpleTest itself, because inherited tests might actually contain several classes/test_methods, therefor the concept of test-datadir makes sense there, which is another reason for avoiding this special treatment.